File: src/comparator.py
from collections import defaultdict
from .models import PriceResult

def choose_cheapest(results: list[PriceResult]) -> dict[str, list[PriceResult]]:
    groups: dict[str, list[PriceResult]] = defaultdict(list)
    for result in results:
        if (
            result.normalized_price is not None
            and result.confidence in {"high", "medium"}
            and result.availability is not False
            and not result.error
        ):
            groups[result.product_id].append(result)
    winners: dict[str, list[PriceResult]] = {}
    for product_id, candidates in groups.items():
        minimum = min(item.normalized_price for item in candidates if item.normalized_price is not None)
        winners[product_id] = [item for item in candidates if item.normalized_price == minimum]
    return winners

File: src/comparator.ts
import type { PriceResult } from "./models.js";

export function chooseCheapest(results: PriceResult[]): Record<string, PriceResult[]> {
  const groups = new Map<string, PriceResult[]>();
  for (const result of results) {
    if (
      result.normalized_price !== null &&
      (result.confidence === "high" || result.confidence === "medium") &&
      result.availability !== false &&
      !result.error
    ) {
      const bucket = groups.get(result.product_id) ?? [];
      bucket.push(result);
      groups.set(result.product_id, bucket);
    }
  }

  const winners: Record<string, PriceResult[]> = {};
  for (const [productId, candidates] of groups) {
    const minimum = candidates.reduce(
      (min, item) => (item.normalized_price!.lt(min) ? item.normalized_price! : min),
      candidates[0]!.normalized_price!,
    );
    winners[productId] = candidates.filter((item) => item.normalized_price!.eq(minimum));
  }
  return winners;
}
